babelsberg/constraintinterpreter.js

```diff
@@ -381,6 +336,110 @@ Object.subclass('Babelsberg', {
+     * requirements laid out in opts.
+     */
+    chooseConstraint: function(constraints, opts, errors) {
+        if (constraints.length === 1)
+            return constraints[0];
+        var constraint = null;
+        var previouslyEnabledConstraints = [];
+        // make sure all constraints are disabled before the comparison
+        constraints.each(function(each) {
+            if (each._enabled)
+                previouslyEnabledConstraints.push(each);
+            each.disable();
+        });
+        for (var i = 0; i < constraints.length; i++) {
+            try {
+                Constraint.current = constraints[i];
+                constraints[i].enable(true);
+                constraints[i].disable();
+            } catch (e) {
+                errors.push(e);
+                constraints[i].disable();
+                constraints[i] = null;
+            } finally {
+                Constraint.current = null;
+            }
+        }
+        var min = Number.MAX_VALUE;
+        var minIndex = -1;
+        for (var i = 0; i < constraints.length; i++) {
+            if (constraints[i] && constraints[i].oComparisonMetrics.time < min) {
+                min = constraints[i].oComparisonMetrics.time;
+                minIndex = i;
+            }
+        }
+        if (minIndex > -1) {
+            constraint = constraints[minIndex];
+            if (opts.logTimings) {
+                console.log('Selected fastest solver:' +
+                            constraint.solver.solverName);
+            }
+        }
+        return constraint;
+    },
+
+    /**
+     * Creates a constraint equivalent to the given function through
+     * Babelsberg#always, and then disables it immediately
+     * @function Babelsberg#once
+     * @public
+     */
+    once: function(opts, func) {
+        var constraint = this.always(opts, func);
+        constraint.disable();
+        return constraint;
+    },
+
+    reevaluateSolverSelection: function(currentConstraint, updatedConstraintVariable) {
+        var currentSolver = currentConstraint.solver;
+        var opts = currentConstraint.originalOpts;
+        var solvers = this.chooseSolvers(opts.solver);
+        solvers = solvers.filter(function(each) { return each !== currentSolver; });
+        solvers = this.filterSolvers(solvers, opts);
+        if (solvers.length < 1)
+            return; // no other solver is qualified to enforce this constraint
+        var errors = [];
+        var constraints = this.createEquivalentConstraints(solvers, opts,
+                                   currentConstraint._predicate, errors);
+        constraints.push(currentConstraint);
+        var constraint = this.chooseConstraint(constraints, opts, errors);
+        if (constraint !== currentConstraint) {
+            currentConstraint.solver = constraint.solver;
+            // yes, constraint does not replace currentConstraint, only its solver
```

Comment:
I think this comment does not say what it should. It says "yes you read that right. Deal with it" rather than "We do this so that... It works because..."